Sample along spline

brisvag opened the pull request:
This PR implements functions to sample an image in a grid along a spline. Main changes:

1. update the spline logic to match the current logic in teamtomo/tomogeom
    - some small changes in the setup
    - added generation of smoothly varying orientations
2. create a generic grid sampler function that takes oriented points and an image, and transforms a regular grid in space based on the points and then uses it to sample from the image
3. a glue function that takes the above and allows to just pass points and an image and get a spline-based grid sampling :)

I split 2 and 3 because this way we can generate oriented points however we like.

Example code that works (no good sample data to share easily, sorry):

```py
import napari
from morphosamplers.sampler import sample_along_spline

v = napari.Viewer()
image_layer = v.open(<some_image>)[0]
points_layer = v.add_points()
```

Then I picked a few points for the spline (on the same slice in this case, but they should work in 3 dimensions):

![image](https://user-images.githubusercontent.com/23482191/197008381-c5050346-17cc-4718-9ff4-febfacf4627c.png)

Then, simply:

```py
result = sample_along_spline(points_layer.data, image_layer.data, grid_shape=(100, 100), n_samples=1000)

v2 = napari.Viewer()
v2.add_image(result)
```

Here's the middle slice of the "straightened" membrane :)

![image](https://user-images.githubusercontent.com/23482191/197008551-de996c4f-d326-4caf-8cec-1a8f4784edf0.png)


I have not yet hardened the code to work with 3d+n dimensions, but it should be relatively easy.

alisterburt commented on src/morphosamplers/sampler.py:
```diff
@@ -0,0 +1,25 @@
+
+
+def sample_with_grid(positions, orientations, image, grid_shape=(10, 10), grid_spacing=1, spline_interpolation_order=3):
+    width = np.linspace(-1, 1, grid_shape[0]) * grid_shape[0] * grid_spacing
+    height = np.linspace(-1, 1, grid_shape[1]) * grid_shape[1] * grid_spacing
+    grid = np.stack(np.meshgrid(width, height, 0), axis=3).reshape(-1, 3)
```
how sure are you that everything is correctly oriented here (e.g. handedness is preserved)
I'm generally not a fan of not being specific with the `indexing` keyword argument of `meshgrid()` 

some comments explaining the logic here will be useful when we come back to this in the future

alisterburt replied:
sorry for the nit but I do think this is super important to get these right from the beginning as these samplers are designed for quantifying morphological stuff

brisvag replied:
Good point. Not sure at all :P This was just a first iteration to get some feedback, will add details and check things now!

alisterburt commented on src/morphosamplers/sampler.py:
```diff
@@ -0,0 +1,25 @@
+from .spline import Spline3D
+
+
+def sample_with_grid(positions, orientations, image, grid_shape=(10, 10), grid_spacing=1, spline_interpolation_order=3):
```
I assume orientations is a scipy `Rotation` - is this correct?

brisvag replied:
yup, I will add some hints and stuff to clarify

alisterburt commented on src/morphosamplers/sampler.py:
```diff
@@ -0,0 +1,25 @@
+    for r in orientations:
+        rotated.append(r.apply(grid))
+    rotated_shifted = rotated + positions.reshape(-1, 1, 3)
+    sampled_image = map_coordinates(image, rotated_shifted.T, order=spline_interpolation_order)
+    return sampled_image.reshape(*grid_shape, -1).T
```
quite a few transposes going on here - same point as above, array handling is very implicit and I find this a bit hard to follow without firing up a console and stepping through

alisterburt commented on src/morphosamplers/sampler.py:
```diff
@@ -0,0 +1,25 @@
+
+
+def sample_along_spline(points, image, n_samples=100, spline_order=3, **kwargs):
+    spline = Spline3D(points=points, order=spline_order)
```
consider taking an already defined spline as input

brisvag replied:
I guess it's useful to have both, right?

alisterburt commented on src/morphosamplers/spline.py:
```diff
@@ -53,140 +49,119 @@ def validate_number_of_points(
+    _rotation_sampler = PrivateAttr(Slerp)
+
+    @validator('points')
+    def _is_n_plus_3d_coordinate_array(cls, v):
```
```suggestion
    def _is_n_by_3_coordinate_array(cls, v):
```

how about the above if you didn't like `is_3d_coordinate_array`? n plus three doesn't make much sense to me

brisvag replied:
Oops, this was aleftover from trying to do `n+3d` handling.

alisterburt commented on src/morphosamplers/utils.py:
```diff
@@ -0,0 +1,36 @@
+import numpy as np
+
+
+def calculate_y_vectors_from_z_vectors(
```
noice

kevinyamauchi replied:
I like this function! It would be nice to add the `Parameters` and `Returns` sections here.

alisterburt commented on tests/test_spline_model.py:
```diff
@@ -121,6 +121,6 @@ def test_invalid_spline_derivatives(derivative_order):
 
     with pytest.raises(ValueError):
-        _ = spline_model._get_equidistance_spline_samples(
+        _ = spline_model._get_equidistant_spline_samples(
```
fyi I renamed this equidistance because I don't think equidistant makes sense for what this spline does

If this was a spline which directly gave you equidistant samples along the original spline then `equidistant_spline` would be a good name - it doesn't do that though, it gives you u values which yield equidistant values from the raw spline -> the spline confers the property of 'equidistance' on u values, hence preference for 'equidistance'

brisvag replied:
I thought this renaming was a merging mistake, that's why I didn't address it, sorry for just overriding without mentioning it :P I see your point, I guess this name would be more correct for the function below which returns equally spaced stuff. I'll roll back!

alisterburt reviewed (commented):
hey @brisvag - functionality is looking really great! few general comments and a few specific comments throughout the PR

- could you separate out the changes to the spline model from the sampling code? There are some changes there I preferred in @kevinyamauchi's version - will be easier to discuss separately if you want them in
- docstrings and variable names would be useful to help understand exactly what's expected
- in sampler.py it feels like the generation of sampling grids should be separated from the actual sampling, both are useful
- the sampling interface might be more flexible if we took in an existing spline as input for sampling rather than constructing a new one in the function - would mean you can get away from duplicating spline logic in sampling code as a bonus

brisvag commented:
I went through the PR and removed a lot of accidental/unnecessary changes, as well as added docstrings and comments... In general, I was a bit trigger happy with this PR, thanks for spending the time to go through it anyways :P

>     * in sampler.py it feels like the generation of sampling grids should be separated from the actual sampling, both are useful

Sure, can split out the functions.

brisvag commented:
Ok, a bunch of updates:

1. removed a lot of changes I inadvertently did to the spline code (sorry about that mess ^^')
2. split out the sampling logic into more atomic functions
3. added docstrings and comments to hopefully explain thing better
4. benefit of point 2. is that now we can sample with 3D grids as well :P

Update of the example above:
```py
import napari
from morphosamplers.sampler import sample_along_spline, Spline3D

v = napari.Viewer()
image_layer = v.open(<some_image>)[0]
points_layer = v.add_points()
```

Some points are picked manually:

![image](https://user-images.githubusercontent.com/23482191/197198777-7211b845-8439-4bba-98aa-d085ba252c27.png)

```py
spline = Spline3D(points=points_layer.data)
sampled = sample_image_along_spline(image_layer.data, spline=spline, n_samples=150, grid_shape=(50, 50))
```

Which makes this (central slice):

![image](https://user-images.githubusercontent.com/23482191/197199067-9990aa6e-3894-4b90-826f-eb6eacdc0c37.png)


However, we can also just sample the points themselves with 3D grids!

```
from morphosamplers.sampler import sample_image_subvolumes

pos = points_layer.data
ori = Rotation.identity(len(pos))
sampled = sample_image_subvolumes(image_layer.data, pos, ori, grid_shape=(50, 50, 50))
```

https://user-images.githubusercontent.com/23482191/197200611-fdafff8f-7f95-4990-93c8-aacaee6aa284.mp4

kevinyamauchi commented on src/morphosamplers/utils.py:
```diff
@@ -0,0 +1,36 @@
+    y = np.empty((len(z), 3))
+
+    # normalise initial y vector so that dot product is the projection
+    if initial_y_vector is None:
+        initial_y_vector = np.array([0, 0, 1])
```
I think it would be more clear to set the default value in the kwargs (e.g., something like, `initial_y_vector`: Union[np.ndarray, Tuple[float, float, float])  = (0, 0, 1)`)

kevinyamauchi commented on src/morphosamplers/spline.py:
```diff
@@ -163,7 +166,7 @@ def _get_equidistance_u(self, separation: float) -> np.ndarray:
 
     def _get_equidistance_spline_samples(
-        self, separation: float, derivative_order: bool = 0
+        self, separation: float, derivative_order: int = 0
```
Nice catch, thanks!

kevinyamauchi commented on src/morphosamplers/spline.py:
```diff
@@ -7,14 +7,17 @@
     points: np.ndarray
     order: conint(ge=1, le=5) = 3
-    _n_spline_samples: int = 10000
+    _n_spline_samples: int = PrivateAttr(10000)
```
Thanks for catching this!

kevinyamauchi commented on src/morphosamplers/utils.py:
```diff
@@ -0,0 +1,36 @@
+
+    This function will return the set of unit vectors perpendicular to the z-vectors which are
+    maximally coaxial to their neighbours. It assumes that z vectors (n, 3) vary
+    smoothly with increasing n.
```
I _think_ this is correct?

```suggestion
    smoothly and monotonically with increasing n.
```

brisvag replied:
No, you can have a "snakey" filament and it will work just fine. What won't "work" is if you have z vector `i` and z vector `i + 1` be basically unrelated, then the results will be just as unrelated... which I guess we could even leave out :P

alisterburt replied:
agree monotonically isn't quite right here for the z vectors, it is right for n but seems unnecessary

kevinyamauchi commented on src/morphosamplers/sampler.py:
```diff
@@ -0,0 +1,112 @@
+):
+    """Extract image planes from a volume following a spline path."""
+    u = np.linspace(0, 1, n_samples)
+    pos = spline.sample_spline(u)
+    ori = spline.sample_spline_orientations(u)
```
I would prefer if the variable names were complete words (i.e., `positions` and `orientations`)

kevinyamauchi reviewed (commented):
Nice work, @brisvag ! Can you please add some tests for the sampler functions and the new `Spline3D` class? Also, can you please add the `Parameters` and `Returns` parts of the docstring to the `sampler.py` functions?

After that, I think this should be good to go! Thanks!

brisvag commented:
Ok, I addressed the comments and added a bunch of docstrings and type hints :)

alisterburt commented on src/morphosamplers/utils.py:
```diff
@@ -0,0 +1,53 @@
+    y = np.empty((len(z), 3))
+
+    # normalise initial y vector so that dot product is the projection
+    if initial_y_vector is None:
+        initial_y_vector = np.array([0, 0, 1])
+    else:
+        initial_y_vector = initial_y_vector / np.linalg.norm(initial_y_vector)
```
```suggestion
    initial_y_vector = np.asarray(initial_y_vector) / np.linalg.norm(initial_y_vector)
```

alisterburt commented on src/morphosamplers/utils.py:
```diff
@@ -0,0 +1,53 @@
+    for i in range(len(y) - 1):
+        yz = np.dot(y[i], z[i + 1])  # projection of y on next z
+        y[i + 1] = y[i] - yz * z[i + 1]  # subtract z component of current y
+        y[i + 1] /= np.linalg.norm(
+            y[i + 1]
+        )  # normalize each iteration to prevent precision issues
```
```suggestion
        y[i + 1] /= np.linalg.norm(y[i + 1])  # prevents precision issues
```

alisterburt replied:
just a tidiness thing

brisvag replied:
I wrapped all this logic in a separate function cause some things were not working and hard to parse and debug.

alisterburt commented on src/morphosamplers/sampler.py:
```diff
@@ -0,0 +1,218 @@
+        Coordinate of points forming the 2D grid.
+    """
+    # create indices for x and y with correct spacing
+    width = np.linspace(-1, 1, grid_shape[0]) * grid_shape[0] * grid_spacing[0]
+    height = np.linspace(-1, 1, grid_shape[1]) * grid_shape[1] * grid_spacing[1]
+    depth = np.linspace(-1, 1, grid_shape[2]) * grid_shape[2] * grid_spacing[2]
```
this yields a grid with values in the range `[-grid_shape[i], grid_shape[i]]` if `grid_spacing[i]==1]`, i.e. length is 2 *  `grid_shape[i]` - is this intended? It seems incorrect but maybe I'm missing something

brisvag replied:
I'm sure there are other ways to frame this, but the way I did it is this:
- `grid_shape`: number of points in the grid along each dimension. So a grid shape of `(5, 5)` spits out a 5 by 5 grid of points
- `grid_spacing`: defines the euclidean distance of the above points.

So if you want a 3D grid to sample subvolumes and you want those subvolumes to be `64x64` pixels with a spacing of `1 * image_pixel_spacing`, you do:

```py
def generate_2D_grid(grid_shape=(64, 64), grid_spacing=(1, 1))
```

brisvag replied:
Another way to do this would be to provide the euclidean size of the grid and the density, which would then become this for the above:
```py
def generate_2D_grid(grid_size=(32, 32), grid_density=(64, 64))
```

I find the first option more intuitive...

alisterburt replied:
Ah okay I'm with you! I misunderstood what was going because of the word shape

maybe rename `grid_spacing` to `grid_point_spacing` to differentiate the two? Or I just accept that I'm an idiot :)

brisvag replied:
Nah it's just that both versions are intuitive depending on how you're approaching :P I don't think adding `point` helps much here to be honest...

kevinyamauchi replied:
I am fine with `grid_spacing` since it is described in the docstring.

brisvag replied:
Actually, you were right that this was wrong. it should be `-0.5` to `0.5`... As it was currently, with a shape of `(2, 2)` and spacing of `(1, 1)`, we ended up with a the right shape, but the points had a distance of 4 rather than 2.

brisvag replied:
Fixed the generation code... it's a bit trickier than I made it at first :P there were 2 main issues:
1. `-1:1` meant we had doubled distances
2. we have to multiply by `shape[axis] - 1`, not just `shape[axis]`, because we have both the edge points

alisterburt commented on src/morphosamplers/sampler.py:
```diff
@@ -0,0 +1,218 @@
+        Coordinate of points forming the 2D grid.
+    """
+    # create indices for x and y with correct spacing
+    width = np.linspace(-1, 1, grid_shape[0]) * grid_shape[0] * grid_spacing[0]
+    height = np.linspace(-1, 1, grid_shape[1]) * grid_shape[1] * grid_spacing[1]
```
same as below, question re total length between start and end of grid

alisterburt commented on src/morphosamplers/sampler.py:
```diff
@@ -0,0 +1,218 @@
+    # convert to coordinate form (x, y, z, 3) (at z = 0)
+    grid = np.stack(np.meshgrid(width, height, 0, indexing="ij"), axis=3)
+    # drop degenerate dim
+    return einops.rearrange(grid, "x y 1 d -> x y d")
```
<3

brisvag replied:
This is where it shines :P but transposing is really too much xD

alisterburt commented on src/morphosamplers/sampler.py:
```diff
@@ -0,0 +1,218 @@
+    """
+    Copy and transform a given sampling grid onto a set of positions and orientations.
+
+    Returns an (n, *grid_shape, 3) grid of sampling points.
```
```suggestion
    Returns a (batch, *grid_shape, 3) batch of coordinate grids for sampling.
```

alisterburt commented on src/morphosamplers/sampler.py:
```diff
@@ -0,0 +1,218 @@
+    grid_shape = sampling_grid.shape
+    grid_coords = sampling_grid.reshape(-1, 3)
+    # apply each orientation to the grid and store the result
+    for ori in orientations:
+        rotated.append(ori.apply(grid_coords))
```
```suggestion
    for orientation in orientations:
        rotated.append(orientation.apply(grid_coords))
```

alisterburt commented on src/morphosamplers/sampler.py:
```diff
@@ -0,0 +1,218 @@
+    return rotated_shifted.reshape(-1, *grid_shape)
+
+
+def sample_volume_with_coordinates(
```
```suggestion
def sample_volume_at_coordinates(
```
this makes more sense to me

alisterburt reviewed (commented):
Looking good! I have some concerns about the grid generation but otherwise this is looking good

consider separating grid generation from sampling code, they seem quite separate but can easily come later

brisvag commented:
Updated based on @alisterburt's comments and added tests for everything!

alisterburt reviewed (commented):
last tiny thing then good to go from me! great work here

brisvag commented:
One question: as I'm working on the surface spline sampling shenanigans, I also needed to generate "`1D`" grids. In the end, they all share the same mechanisms, so I was thinking of putting it all together like this:

```py
def generate_3D_grid(
    grid_shape: Tuple[int, int, int] = (10, 10, 10),
    grid_spacing: Tuple[float, float, float] = (1, 1, 1),
    squeeze: bool = True,
) -> np.ndarray:
    """
    Generate a 3D sampling grid with specified shape and spacing.

    The grid generated is centered on the origin, has shape (*grid_shape, 3)
    and spacing grid_spacing between neighboring points.

    Parameters
    ----------
    grid_shape : Tuple[int, int, int]
        Shape of the 3D grid.
    grid_spacing : Tuple[float, float, float]
        Spacing between points in the sampling grid.
    squeeze: bool
        Discard degenerate dimensions in the output.

    Returns
    -------
    np.ndarray
        Coordinate of points forming the 2D grid.
    """
    # create indices for x and y with correct spacing
    w, h, d = grid_shape
    dw, dh, dd = grid_spacing
    if w != 1:
        width = np.linspace(-1, 1, w) * w * dw
    else:
        w = 0
    if h != 1:
        height = np.linspace(-1, 1, h) * h * dh
    else:
        h = 0
    if d != 1:
        depth = np.linspace(-1, 1, d) * d * dd
    else:
        d = 0
    # convert to stack form (x, y, z, 3)
    grid = np.stack(np.meshgrid(width, height, depth, indexing="ij"), axis=3)
    return grid
```

Which then would mean that the current `2d` version is done with:
```py
generate_3D_grid(grid_shape=(*grid_shape, 1), grid_spacing=(*grid_spacing, 1))
```

and if I want a `1D` version which goes along Z, i can do:
```py
    grid = generate_3D_grid(grid_shape=(0, 0, sampling_thickness), grid_spacing=(0, 0, sampling_spacing))
```

If one wants to keep degenerate dimensions (for example for sampling subvolumes with a single pixel in one dimensions but without them being merged together) they can pass `squeeze=False`, but I don't expect that to be very used.

Does it make sense or is it a bit unwieldy?

alisterburt commented:
makes sense to me! I don't mind either way but don't mind the code duplication particularly - I find it easier if I don't have to flit around so much when the logic isn't crazy complicated, would probably keep them separate but up to you

kevinyamauchi commented on src/morphosamplers/sampler.py:
```diff
@@ -0,0 +1,218 @@
+    """
+    Generate a 2D sampling grid with specified shape and spacing.
+
+    The grid generated is centered on the origin, lying on the xy plane,
```
I think we should specify the normal vector (`[1, 0, 0]`) as xy plane will mean different things to different people.

kevinyamauchi commented on src/morphosamplers/sampler.py:
```diff
@@ -0,0 +1,218 @@
+    """
+    Generate a 3D sampling grid with specified shape and spacing.
+
+    The grid generated is centered on the origin, has shape (*grid_shape, 3)
```
I personally don't find `(*grid_shape, 3)` clear notation. Perhaps something like shape `(m * n * p, 3)` for gride_shape `(m, n, p)` (assuming that's correct)?

(same for the 2D grid function)

kevinyamauchi commented on tests/test_sampling.py:
```diff
@@ -0,0 +1,73 @@
+
+
+def test_generate_2D_grid():
+    shape = (2, 5)
```
Nit: it would be nice to test the spacing (same for 3d grid)

kevinyamauchi replied:
Sorry to be pedantic, but as I read it, these tests to check that the spacing between each point is correct (e.g., unequal spacing, but correct end points would pass). That being said, I as mentioned above, this is a nitpick, so I'm fine leaving it as-is if you prefer.

brisvag replied:
Ah I see what you mean. Will update!

kevinyamauchi commented on src/morphosamplers/sampler.py:
```diff
@@ -0,0 +1,218 @@
+    Parameters
+    ----------
+    grid_shape : Tuple[int, int, int]
+        Shape of the 3D grid.
```
```suggestion
        The number of grid points along each axis.
```

kevinyamauchi commented on src/morphosamplers/sampler.py:
```diff
@@ -0,0 +1,218 @@
+    Parameters
+    ----------
+    grid_shape : Tuple[int, int]
+        Shape of the 2D grid.
```
```suggestion
        The number of grid points along each axis.
```

kevinyamauchi reviewed (commented):
Thanks for all of the updates, @brisvag ! I think this is pretty much good to go once we settle your question regarding the grid API and the tests are passing.

> One question: as I'm working on the surface spline sampling shenanigans, I also needed to generate "1D" grids. In the end, they all share the same mechanisms, so I was thinking of putting it all together like this:
> ...
> Does it make sense or is it a bit unwieldy?

I am okay with this if you keep convenience functions `generate_2d_grid()` and `generate_3d_grid()` (they can call the the `generate_3d_grid()` function in the background). My concern is that I don't think that is going to be a very intuitive API for most people, so I think we should keep the simple convenience functions.

brisvag commented:
@alisterburt @kevinyamauchi I think this is ready to go in!

alisterburt commented on src/morphosamplers/sampler.py:
```diff
@@ -0,0 +1,223 @@
+from .spline import Spline3D
+
+
+def generate_3D_grid(
```
```suggestion
def generate_3d_grid(
```

alisterburt replied:
feels more conventional

alisterburt commented on src/morphosamplers/sampler.py:
```diff
@@ -0,0 +1,223 @@
+    """
+    Generate a 3D sampling grid with specified shape and spacing.
+
+    The grid generated is centered on the origin, has shape (m, n, p, 3) for
+    grid_shape (m, n, p), and spacing grid_spacing between neighboring points.
```
feel free to include or not, feels more obvious to me this way

alisterburt replied:
don't think so, h and w for height and width of 2d images, third dimension is called depth and is added in front

alisterburt commented on src/morphosamplers/sampler.py:
```diff
@@ -0,0 +1,223 @@
+    np.ndarray
+        Coordinate of points forming the 3D grid.
+    """
+    # create indices for x and y with correct spacing
```
outdated? should it be xyz?

alisterburt commented on src/morphosamplers/sampler.py:
```diff
@@ -0,0 +1,223 @@
+    """
+    # create indices for x and y with correct spacing
+    coords = []
+    for dim in range(3):
+        ln = grid_shape[dim]
+        spacing = grid_spacing[dim]
+        if ln != 1:
+            coord = np.linspace(-0.5, 0.5, ln) * (ln - 1) * spacing
+        else:
+            coord = 0
+        coords.append(coord)
+    # convert to stack form (x, y, z, 3)
+    grid = np.stack(np.meshgrid(*coords, indexing="ij"), axis=3)
```
This could be replaced with below which feels a little clearer - I like to keep the names I use for image dimensions separate from the names I use for coordinates which index into those dimensions

```suggestion
    grid = np.indices(grid_shape)
    grid = einops.rearrange(grid, 'zyx d h w -> d h w zyx')
    grid -= np.array(grid_shape) / 2
    grid *= spacing
```

```

brisvag replied:
Awesome! But it's off by 0.5, will fix.

alisterburt commented on src/morphosamplers/sampler.py:
```diff
@@ -0,0 +1,223 @@
+    np.ndarray
+        Coordinate of points forming the 2D grid.
+    """
+    return generate_3D_grid(grid_shape=(*grid_shape, 1), grid_spacing=(*grid_spacing, 1))
```
could also squeeze with einops ;)

alisterburt commented on src/morphosamplers/sampler.py:
```diff
@@ -0,0 +1,223 @@
+    grid_shape = sampling_grid.shape
+    grid_coords = sampling_grid.reshape(-1, 3)
+    # apply each orientation to the grid and store the result
+    for orientation in orientations:
+        rotated.append(orientation.apply(grid_coords))
```
using a `Rotation` object here means we should be careful to be very explicit about the expected order of the coordinates in the sampling grid

brisvag replied:
I guess you're referring to the `dwh` issue above, or am I misunderstanding?

alisterburt commented on src/morphosamplers/sampler.py:
```diff
@@ -0,0 +1,223 @@
+    )
+    # reshape back (need to invert due to previous transposition)
+    # and retranspose to get batch back to the 0th dimension
+    return np.swapaxes(sampled_volume.reshape(*grid_shape, batch), -1, 0)
```
suggestion, not mandatory for sure

alisterburt replied:
`'xyz ... -> ... xyz'` would move the first axis to the end

brisvag replied:
sweet!

alisterburt commented on src/morphosamplers/spline.py:
```diff
@@ -190,3 +196,62 @@ def _get_equidistance_spline_samples(
+
+
+class Spline3D(NDimensionalSpline):
+    """3D spline model with interpolation of orientations along the spline.
```
```suggestion
    """3D spline model with a consistent local coordinate system.
```

alisterburt reviewed (commented):
Looking really really great - suggestions are minor in most cases feel free to either address or not

The only one I'm a bit concerned about is the interaction between `Rotation` objects and the coordinate grids being a little fuzzy, need to be very clear about expected ordering of coordinates in the grids

kevinyamauchi reviewed (approved):
Thanks, @brisvag ! This looks good to me. Thank you for the additional tests. I think this is good to go once you decide which of @alisterburt 's suggestions to incorporate.

brisvag commented:
I either adrressed or responded to the above comments, should be basically done!

alisterburt commented:
Looks like some minor import error here then we're good to go, `einops` -> `einops.rearrange` should sort it!

brisvag commented:
oopsie!

kevinyamauchi commented:
Test are now passing, so I will merge. Thanks again, @brisvag !
